2-WebApp-graph-user/2-1-Call-MSGraph/Controllers/HomeController.cs

```diff
@@ -18,27 +23,55 @@ public class HomeController : Controller
 
         private readonly GraphServiceClient _graphServiceClient;
 
+        private readonly MicrosoftIdentityConsentAndConditionalAccessHandler _consentHandler;
+
+        private string[] _graphScopes = new[] { "user.read" };
+
         public HomeController(ILogger<HomeController> logger,
-                          GraphServiceClient graphServiceClient)
+                            IConfiguration configuration,
+                            GraphServiceClient graphServiceClient,
+                            MicrosoftIdentityConsentAndConditionalAccessHandler consentHandler)
         {
             _logger = logger;
             _graphServiceClient = graphServiceClient;
+            this._consentHandler = consentHandler;
+
+            _graphScopes = configuration.GetValue<string>("DownstreamApi:Scopes")?.Split(' ');
         }
 
         [AuthorizeForScopes(ScopeKeySection = "DownstreamApi:Scopes")]
-        public async Task<IActionResult> Index()
+        public IActionResult Index()
         {
-            var user = await _graphServiceClient.Me.Request().GetAsync();
-            ViewData["ApiResult"] = user.DisplayName;
-
             return View();
         }
 
         [AuthorizeForScopes(ScopeKeySection = "DownstreamApi:Scopes")]
         public async Task<IActionResult> Profile()
         {
-            var me = await _graphServiceClient.Me.Request().GetAsync();
-            ViewData["Me"] = me;
+            User currentUser = null;
+
+            try
+            {
+                currentUser = await _graphServiceClient.Me.Request().GetAsync();
+            }
+            catch (System.Exception ex) // Catch CAE exception from Graph SDK

```

Comment:
- Why not catch a `ServiceException` directly ?
- Why use `Trim` given the test uses `Contains`?
- I would recommend using the when syntax here;
- Would there be an AADSTS error message (in case the message from EVO is localized ....)  ?

```CSharp
catch (ServiceException svcex) when (svcex.Message.Contains("Continuous access evaluation resulted in claims challenge"))
{
    try
    {
        Console.WriteLine($"{svcex}");
        var claimChallenge = AuthenticationHeaderHelper.ExtractClaimChallengeFromHttpHeader(svcex.ResponseHeaders);
        _consentHandler.ChallengeUser(_graphScopes, claimChallenge);
        return new EmptyResult();
      }
      catch (Exception ex2)
      {
            _consentHandler.HandleException(ex2);
      }
}

```